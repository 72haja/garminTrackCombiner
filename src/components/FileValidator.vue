<script setup>
import { computed, watch, ref } from 'vue';

const errorMessage = ref("")
const fileActivityTypes = ref([])

const props = defineProps({
  files: {
    type: Array,
    default: () => ([])
  }
})

const emit = defineEmits(['error-exists'])

watch(() => props.files, async () => {
  getFileActivityTypes();
  checkIfAllDocsHaveSameActivity();
},
  {
    immediate: true
  }
)

const errorMessageLength = computed(() => errorMessage.value?.length || 0)

function getFileActivityTypes () {
  if (!props.files) return []

  fileActivityTypes.value = props.files.map((file) => {
    const sportType = file.content.getElementsByTagName("Activity")[0].getAttribute("Sport")

    return {
      name: file.name,
      sportType: sportType,
    }
  })
}

function checkIfAllDocsHaveSameActivity () {
  emit("error-exists", false)
  const firstActivity = fileActivityTypes.value[0].sportType
  fileActivityTypes.value.forEach((doc) => {
    const docActivity = doc.sportType
    if (docActivity !== firstActivity) {
      errorMessage.value = "Alle XML-Dateien müssen die gleiche Aktivität haben."
      emit("error-exists", true)
    }
  })
}
</script>

<template>
  <div class="pa-4 custom-list">
    <v-list-item v-for="(file, index) in fileActivityTypes"
                 :key="index"
                 elevation="2">
      <div class="d-flex justify-space-between align-center">
        <span>
          {{ file.name }}
        </span>
        <span>
          {{ file.sportType }}
        </span>
      </div>
    </v-list-item>
  </div>
  <v-alert v-if="errorMessageLength > 0"
           type="error"
           dismissible
           class="mb-4">
    {{ errorMessage }}
  </v-alert>

</template>
<style lang='scss' scoped>
.custom-list {
  display: flex;
  flex-direction: column;
  gap: 5px;
}
</style>
