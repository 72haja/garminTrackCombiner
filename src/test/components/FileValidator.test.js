import { describe, expect, vi } from 'vitest'
import { mount, shallowMount } from '@vue/test-utils'

import { createVuetify } from 'vuetify'

import testFile1 from '../../../public/mocks/testFile1.tcx?raw'
import testFile2 from '../../../public/mocks/testFile2.tcx?raw'

import FileValidator from '../../components/FileValidator.vue'
import { nextTick } from 'vue'

function getXmlDoc (content) {
  const parser = new DOMParser();
  return parser.parseFromString(content, "text/xml");
}

const testFile1Xml = getXmlDoc(testFile1)
const testFile2Xml = getXmlDoc(testFile2)

describe('mount App', () => {
  const vuetify = createVuetify()

  test('mount component', async () => {
    expect(FileValidator).toBeTruthy()

    const wrapper = mount(FileValidator, {
      props: {
        files: [
          {
            name: "testFile1Xml",
            content: testFile1Xml,
          },
          {
            name: "testFile2Xml",
            content: testFile2Xml,
          },
        ],
      },
      global: {
        plugins: [vuetify]
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('test computed', () => {
  const vuetify = createVuetify()
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(FileValidator, {
      props: {
        files: [
          {
            name: "testFile1Xml",
            content: testFile1Xml,
          },
          {
            name: "testFile2Xml",
            content: testFile2Xml,
          },
        ],
      },
      global: {
        plugins: [vuetify]
      }
    });
  });

  describe('errorMessageLength', () => {
    test('errorMessageLength - 0 - does not exist', async () => {
      wrapper.vm.errorMessage = null

      expect(wrapper.vm.errorMessageLength).toBe(0)
    })

    test('errorMessageLength - 1 - exists', async () => {
      wrapper.vm.errorMessage = "error"

      expect(wrapper.vm.errorMessageLength).toBeGreaterThan(0)
    })
  });
})

describe('test function', () => {
  const vuetify = createVuetify()
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(FileValidator, {
      props: {
        files: [
          {
            name: "testFile1Xml",
            content: testFile1Xml,
          },
          {
            name: "testFile2Xml",
            content: testFile2Xml,
          },
        ],
      },
      global: {
        plugins: [vuetify]
      }
    });
  });

  describe('getFileActivityTypes', () => {
    test('getFileActivityTypes - activity types', async () => {
      wrapper.vm.getFileActivityTypes()

      await nextTick()

      expect(
        wrapper.vm.fileActivityTypes
      ).toEqual([
        {
          name: "testFile1Xml",
          sportType: "Biking",
        },
        {
          name: "testFile2Xml",
          sportType: "Biking",
        }
      ])
    })

    test('getFileActivityTypes - no activity types', async () => {
      await wrapper.setProps({
        files: null
      });

      wrapper.vm.getFileActivityTypes()

      await nextTick()

      expect(
        wrapper.vm.fileActivityTypes
      ).toEqual([])
    })
  })

  describe('checkIfAllDocsHaveSameActivity', () => {
    test('checkIfAllDocsHaveSameActivity - truthy', async () => {
      wrapper.vm.checkIfAllDocsHaveSameActivity()
      
      await nextTick()
      
      expect(wrapper.emitted('error-exists')).toBeTruthy();
      expect(wrapper.emitted('error-exists')).toContainEqual([false]);
      expect(wrapper.emitted('error-exists')).not.toContainEqual([true]);
    })
    
    test('checkIfAllDocsHaveSameActivity - falsy', async () => {
      wrapper.vm.fileActivityTypes = [
        {
          name: "testFile1Xml",
          sportType: "Biking",
        },
        {
          name: "testFile2Xml",
          sportType: "Running",
        }
      ]
      
      wrapper.vm.checkIfAllDocsHaveSameActivity()
      
      await nextTick()
      
      expect(wrapper.emitted('error-exists')).toBeTruthy();
      expect(wrapper.emitted('error-exists')).toContainEqual([false]);
      expect(wrapper.emitted('error-exists')).toContainEqual([true]);
    })
  })
})

