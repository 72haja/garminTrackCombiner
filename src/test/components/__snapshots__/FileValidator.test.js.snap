// Vitest Snapshot v1

exports[`mount App > mount component 1`] = `
"<div class=\\"pa-4 custom-list\\" data-v-45a53a49=\\"\\">
  <div class=\\"v-list-item v-theme--light v-list-item--density-default elevation-2 v-list-item--variant-text\\" data-v-45a53a49=\\"\\">
    <!----><span class=\\"v-list-item__underlay\\"></span>
    <!---->
    <div class=\\"v-list-item__content\\">
      <!---->
      <!---->
      <div class=\\"d-flex justify-space-between align-center\\" data-v-45a53a49=\\"\\"><span data-v-45a53a49=\\"\\">testFile1Xml</span><span data-v-45a53a49=\\"\\">Biking</span></div>
    </div>
    <!---->
  </div>
  <div class=\\"v-list-item v-theme--light v-list-item--density-default elevation-2 v-list-item--variant-text\\" data-v-45a53a49=\\"\\">
    <!----><span class=\\"v-list-item__underlay\\"></span>
    <!---->
    <div class=\\"v-list-item__content\\">
      <!---->
      <!---->
      <div class=\\"d-flex justify-space-between align-center\\" data-v-45a53a49=\\"\\"><span data-v-45a53a49=\\"\\">testFile2Xml</span><span data-v-45a53a49=\\"\\">Biking</span></div>
    </div>
    <!---->
  </div>
</div>
<!--v-if-->"
`;
